pecos/distributed/xmc/xlinear/READMD.md

```diff
@@ -0,0 +1,172 @@
+You should receive N "Hello, World!" messages from each machine of the cluster.
+
+## Getting started
+
+### Basic Command-line Usage
+On the main node:
+```bash
+mpiexec -f hostfile -n ${NUM_MACHINE} python3 -m pecos.distributed.xmc.xlinear.train -x ${X_PATH} -y ${Y_PATH} -m ${MODEL_DIR}
+```
+where:
+* `NUM_MACHINE` is the number of machines used for distributed training, should <= number of machines in the cluster.
+* `X_PATH` is the paths to the CSR npz or Row-majored npy files of the training feature matrices with shape `(N, d)`.
+* `Y_PATH` is the paths to the CSR npz files of the training label matrices with shape `(N, L)`.
+* `MODEL_DIR` is the path to the model folder where the trained model will be saved to, will be created if not exist.
+
+All the paths: `X_PATH`, `Y_PATH` and `MODEL_DIR` should be accessible by all machines on a shared network disk.
+
+For detailed parameters list, please refer to:
+```bash
+python3 -m pecos.distributed.xmc.xlinear.train --help
+```
+
+After training the model, you can use [`pecos.xmc.xlinear.predict`](pecos/xmc/xlinear/README.md#basic-command-line-usage) to do inferences.
+
+The model trained by distributed `pecos.distributed.xmc.xlinear.train` is almost identical to the model trained by single-box `pecos.xmc.xlinear.train`, if not considering the randomness introduced by sub-models. Accordingly, the inference results from distributed and single-box models should also be very similar.
+
+### Example: Distributed XLinear Model Training on eurlex-4k Data
+Make sure you have setup all above hardware, software and workspace prerequisites.
+
+Prepare `eurlex-4k` data:
+```bash
+cd <SHARED-NETWORK-DISK-PATH>/pecos-workspace
+
+wget https://archive.org/download/pecos-dataset/xmc-base/eurlex-4k.tar.gz
+tar -zxvf eurlex-4k.tar.gz
+```
+
+Train:
+```
+mpiexec -f hostfile \
+python3 -m pecos.distributed.xmc.xlinear.train \
+-x ./xmc-base/eurlex-4k/tfidf-attnxml/X.trn.npz \
+-y ./xmc-base/eurlex-4k/Y.trn.npz \
+--nr-splits 2 -b 50 -k 100 -m eurlex_4k_model \
+--min-n-sub-tree 16 -t 0.1 \
+--meta-label-embedding-method pii \
+--sub-label-embedding-method pifa \
+--verbose-level 3
+```
+
+Predict:
+```
+python3 -m pecos.xmc.xlinear.predict \
+-x ./xmc-base/eurlex-4k/tfidf-attnxml/X.tst.npz \
+-y ./xmc-base/eurlex-4k/Y.tst.npz \
+-m ./eurlex_4k_model
+```
+
+## Appendix: Distributed Efficiencies Q&A
+Extra distributed training parameters are available for better time and memory efficiency.
+
+**Q: What should I do if distributed model training is too slow, or consume too much memory on main node?**
+
+A: Use `PII` embedding for meta-tree label embedding to accelerate the clustering:
+```bash
+mpiexec -f hostfile -n ${NUM_MACHINE} \
+python3 -m pecos.distributed.xmc.xlinear.train \
+-x ${X_PATH} -y ${Y_PATH} -m ${MODEL_DIR} \
+--meta-label-embedding-method pii
+```
+This would produce slightly different models.
+
+**Q: What should I do if my model is too big, and MPI report buffer exceed limit error in training?**
```

Comment:
"model is too big" is too vague an issue. I'd recommend to paste the actual MPIBufferError here since the solution is particular for that error.